sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,388 @@
+      bitfield_bit32_write(0, SPI_HOST_CONTROL_SPIEN_BIT, enable));
+}
+
+dif_result_t dif_spi_host_configure(const dif_spi_host_t *spi_host,
+                                    dif_spi_host_config_t config) {
+  if (spi_host == NULL) {
+    return kDifBadArg;
+  }
+  if (config.peripheral_clock_freq_hz == 0 || config.spi_clock == 0) {
+    return kDifBadArg;
+  }
+
+  uint32_t divider = (config.peripheral_clock_freq_hz / config.spi_clock) - 1;
+  if (divider & ~SPI_HOST_CONFIGOPTS_CLKDIV_0_MASK) {
+    return kDifBadArg;
+  }
+
+  spi_host_reset(spi_host);
+  uint32_t reg = 0;
+  reg =
+      bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CLKDIV_0_FIELD, divider);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNIDLE_0_FIELD,
+                               config.chip_select.idle);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNTRAIL_0_FIELD,
+                               config.chip_select.trail);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNLEAD_0_FIELD,
+                               config.chip_select.lead);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT,
+                             config.full_cycle);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPHA_0_BIT, config.cpha);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPOL_0_BIT, config.cpol);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_CONFIGOPTS_REG_OFFSET, reg);
+  spi_host_enable(spi_host, true);
+  return kDifOk;
+}
+
+static void wait_ready(const dif_spi_host_t *spi_host) {
+  bool ready;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    ready = bitfield_bit32_read(reg, SPI_HOST_STATUS_READY_BIT);
+  } while (!ready);
+}
+
+static void wait_tx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == 0xFF);
+}
+
+static void wait_rx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (rxqd == 0);
+}
+
+static inline void tx_fifo_write8(const dif_spi_host_t *spi_host,
+                                  uintptr_t srcaddr) {
+  uint8_t *src = (uint8_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
```

Comment:
Isn't this what `mmio_memcpy` is supposed to be fore?

Reply:
My understanding is that `mmio_memcpy` works like `memcpy`, but for IO.  That is, it copies from src[0..len] to dst[0..len].

The SPI host FIFOs are managed through a single register.  Each write to the TXFIFO register inserts another word (or byte) into the FIFO.  Each read from the RXFIFO register drains a word from the FIFO.

Reply:
Oh, I see. Yup, that wouldn't work. I feel it's sufficiently interesting that maybe it's worth factoring out into mmio.h too...?

Reply:
I think that depends on how many other IPs need a `mmio_memcpy_{to,from}_a_single_register` and whether or not the register supports byte-access (TXFIFO does) or not (RXFIFO does not).

Reply:
Shrug.

Reply:
I think it is a very useful utility on mmio.h, DIFs like Kmac, Hmac and Uart can use it.

Reply:
Each of those IP blocks has a different register/mechanism for knowing if there is room left in their FIFOs, as does this one.  

Furthermore, the UART allows only byte writes, while the hmac and kmac register interfaces permit byte and word access.